File: Day177_2025/Q177_Leetcode.java
class Q177_Leetcode {
    public int longestSubsequence(String s, int k) {
        int sm = 0;
        int cnt = 0;
        int bits = 32 - Integer.numberOfLeadingZeros(k);

        int n = s.length();
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(n - 1 - i);
            if (ch == '1') {
                if (i < bits && sm + (1 << i) <= k) {
                    sm += 1 << i;
                    cnt++;
                }
            } else {
                cnt++;
            }
        }
        return cnt;
    }
}

File: Day177_2025/Q177_Leetcode.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        sm = 0
        cnt = 0
        bits = k.bit_length()

        n = len(s)
        for i in range(n):
            ch = s[n - 1 - i]
            if ch == '1':
                if i < bits and sm + (1 << i) <= k:
                    sm += 1 << i
                    cnt += 1
            else:
                cnt += 1

        return cnt
